crypto/poly1305/poly1305_meth.c

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+#include <openssl/evp.h>
+#include "internal/evp_int.h"
+#include "internal/poly1305.h"
+#include "internal/cryptlib.h"
+#include "poly1305_local.h"
+
+/* typedef EVP_MAC_IMPL */
+struct evp_mac_impl_st {
+    POLY1305 *ctx;               /* poly1305 context */
+};
+
+static EVP_MAC_IMPL *poly1305_new(void)
+{
+    EVP_MAC_IMPL *ctx;
+
+    if ((ctx = OPENSSL_zalloc(sizeof(*ctx))) == NULL
+            || (ctx->ctx = OPENSSL_zalloc(sizeof(POLY1305))) == NULL) {
+        OPENSSL_free(ctx);
+        return 0;
+    }
+    return ctx;
+}
+
+static void poly1305_free(EVP_MAC_IMPL *ctx)
+{
+    if (ctx != NULL) {
+        OPENSSL_free(ctx->ctx);
+        OPENSSL_free(ctx);
+    }
+}
+
+static int poly1305_copy(EVP_MAC_IMPL *dst, EVP_MAC_IMPL *src)
+{
+    if ((dst = poly1305_new()) == NULL
+            || !memcpy(dst->ctx, src->ctx, sizeof(*dst->ctx))) {
+        poly1305_free(dst);
+        return 0;
+    }
```

Comment:
This is incorrect.  All you need is this:

```
    memcpy(dst->ctx, src->ctx, sizeof(*dst->ctx));
```

No need to check memcpy's return value, it always returns the destination.

Reply:
With that change, I'm sure the CIs are going to be much happier...

Reply:
Actually I made a stupid mistake for assuming the dst->ctx is NULL which is in fact not...

Reply:
So just a simple `*dst->ctx = *src->ctx` would enough for this `copy`